lib/proxy/router.go

```diff
@@ -313,9 +314,51 @@ func (r *Router) DialHost(ctx context.Context, clientSrcAddr, clientDstAddr net.
 		return nil, trace.Wrap(err)
 	}
 
+	// SSH connection MUST start with "SSH-2.0" bytes according to https://datatracker.ietf.org/doc/html/rfc4253#section-4.2
+	conn = newCheckedPrefixWriter(conn, []byte("SSH-2.0"))
 	return NewProxiedMetricConn(conn), trace.Wrap(err)
 }
 
+// checkedPrefixWriter checks that first data written into it has the specified prefix.
+type checkedPrefixWriter struct {
+	net.Conn
+
+	requiredPrefix  []byte
+	requiredPointer int
+}
+
+func newCheckedPrefixWriter(conn net.Conn, requiredPrefix []byte) *checkedPrefixWriter {
+	return &checkedPrefixWriter{
+		Conn:           conn,
+		requiredPrefix: requiredPrefix,
+	}
+}
+
+func (c *checkedPrefixWriter) Write(p []byte) (int, error) {
+	// If pointer reached end of required prefix the check is done
+	if len(c.requiredPrefix) == c.requiredPointer {
+		return c.Conn.Write(p)
+	}
+
+	// Decide which is smaller, provided data or remaining portion of the require prefix
+	small := c.requiredPrefix[c.requiredPointer:]
+	big := p
+	if len(small) > len(big) {
+		big = small
+		small = p
+	}
+
+	if !bytes.HasPrefix(big, small) {
+		return 0, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
+	}
+	n, err := c.Conn.Write(p)
+	if err == nil {
+		// Advance pointer by confirmed portion (small) of the prefix.
+		c.requiredPointer += len(small)
+	}
```

Comment:
`io.Writer` is one of those weird APIs where the return value is meaningful even in case of error.
```suggestion
	c.requiredPrefix = c.requiredPrefix[min(n, len(c.requiredPrefix)):]
```